modules/vectorfieldvisualization/integrallinetracer.h

```diff
@@ -32,34 +32,281 @@
+        size_t stepsBWD = 0;
 
-    IntegralLineTracer(const IntegralLineProperties &properties);
-    virtual ~IntegralLineTracer();
+        if (both) {
+            stepsBWD = steps_ / 2;
+            stepsFWD = steps_ - stepsBWD;
+        } else if (fwd) {
+            line.setBackwardTerminationReason(IntegralLine::TerminationReason::StartPoint);
+            stepsFWD = steps_;
+        } else if (bwd) {
+            line.setForwardTerminationReason(IntegralLine::TerminationReason::StartPoint);
+            stepsBWD = steps_;
+        }
 
-    IntegralLineProperties::IntegrationScheme getIntegrationScheme() const;
-    void setIntegrationScheme(IntegralLineProperties::IntegrationScheme scheme);
+        stepsBWD++;  // for adjendency info
+        stepsFWD++;
+
+        line.getPositions().reserve(steps_ + 2);
+        line.getMetaData<dvec3>("velocity", true).reserve(steps_ + 2);
+
+        if (TimeDependent) {
+            line.getMetaData<double>("timestamp", true).reserve(steps_ + 2);
+        }
+
+        for (auto &m : metaSamplers_) {
+            line.getMetaData<typename Sampler::ReturnType>(m.first, true).reserve(steps_ + 2);
+        }
+
+        if (!addPoint(line, p)) {
+            return res;  // Zero velocity at seed point
+        }
+
+        integrate(stepsBWD, p, line, false);
+
+        if (!line.getPositions().empty()) {
+            line.reverse();
+            res.seedIndex = line.getPositions().size() - 1;
+        }
+
+        integrate(stepsFWD, p, line, true);
+
+        return res;
+    }
+
+    void addMetaDataSampler(const std::string &name, std::shared_ptr<const Sampler> sampler) {
+        metaSamplers_[name] = sampler;
+    }
+
+    const DataHomogenouSpatialMatrixrix &getSeedTransformationMatrix() const {
+        return seedTransformation_;
+    }
+
+    void setTransformOutputToWorldSpace(bool transform) {
+        transformOutputToWorldSpace_ = transform;
+    }
+    bool isTransformingOutputToWorldSpace() const { return transformOutputToWorldSpace_; }
+
+private:
+    bool addPoint(IntegralLine &line, SpatialVector pos) {
+        auto worldVelocty = sampler_->sample(pos);
+
+        if (glm::length(worldVelocty) < std::numeric_limits<double>::epsilon()) {
+            return false;
+        }
+
+        if (transformOutputToWorldSpace_) {
+            SpatialVector worldPos =
+                detail::seedTransform<Sampler, DataVector, DataHomogenousVector>(
+                    typename std::integral_constant<bool, TimeDependent>::type(), toWorld_, pos);
+
+            line.getPositions().emplace_back(util::glm_convert<dvec3>(worldPos));
+        } else {
+            line.getPositions().emplace_back(util::glm_convert<dvec3>(pos));
+        }
+
+        line.getMetaData<dvec3>("velocity").emplace_back(util::glm_convert<dvec3>(worldVelocty));
+
+        if (TimeDependent) {
+            line.getMetaData<double>("timestamp").emplace_back(pos[Sampler::SpatialDimensions - 1]);
```

Comment:
All of these getters should probably be lifted outside of the for loop in integrate...